Specify how to propagate head sampling probability (#168)
open-telemetry/oteps: text/trace/0168-sampling-propagation.md

`tracestate` with an `otelprob` vendor tag.

## Explanation

Two pieces of information are needed to convey consistent head trace
sampling probability:

1. The head trace sampling probability
2. Source of consistent sampling decisions.

This proposal uses one byte of information for each of these.

### Probability value

To limit the cost of this extension and for statistical reasons
documented below, we propose to limit head trace sampling probability
to powers of two.  This limits the available head trace sampling
probabilities to 1/2, 1/4, 1/8, and so on.  We can compactly encode
these probabilities as small integer values using the base-2 logarithm
of the adjusted count (i.e., inverse probability).

For example, the probability value 2 corresponds with 1-in-4 sampling,
the probability value 10 corresponds with 1-in-1024 sampling.  Using
one byte of information we can convey sampling rates as small as 2^-255.

### Random value

With head trace sampling probabilities limited to powers of two, the
amount of randomness needed per trace context is limited.  A
consistent sampling decision is accomplished by propagating a

> geometrically distributed random variable with shape parameter `1/2`,

Reviewer: > geometrically distributed random variable

define it?

> requiring only two bits of randomness on average per trace.  See

Reviewer: not following this

Author: I've rewritten this passage, hoping to clarify.
```
A
consistent sampling decision is accomplished by propagating a specific
random variable.  The random variable is a described by a discrete
geometric distribution having shape parameter `1/2`, listed below:

| Value | Probability |
| ----- | ----------- |
| 0 | 1/2 |
| 1 | 1/4 |
| 2 | 1/8 |
| 3 | 1/16 |
| 4 | 1/32 |
| ... | ... |
| N | 1/(2^(N+1)) |

See [Estimation from Partially Sampled Distributed
Traces](https://arxiv.org/pdf/2107.07703.pdf) section 2.8 for a
detailed explanation.

Such a random variable `r` can be generated using the following
pseudocode:

```
r := 0
for {
  if nextRandomBit() {
    break // The expected value of r is 2
  }
  r++
}
```
```

[Estimation from Partially Sampled Distributed
Traces](https://arxiv.org/pdf/2107.07703.pdf) section 2.8 for a
detailed explanation.

Such a random variable `r` can be generated using the following
pseudocode:

```
r := 0
for {
  if nextRandomBit() {
    break // The expected value of r is 2
  }
  r++
}
```

> This can be computed from a stream of random bits as the number of

Reviewer: if the intent is to use traceIdRatio sampler, this section is completely ignoring the randomness already present in the trace id

Author: Yes. This has been discussed in past meetings of the Sampling SIG, which led to this issue: https://github.com/open-telemetry/opentelemetry-specification/issues/1826

The OTel specification says very little about how random a TraceID should be. @carlosalberto wrote:

> One detail to be aware of here is that some vendors may be providing their existing ids, so we have to make sure this works in all scenarios.

if we are not able to dictate the degree of actual randomness, we may not be able to use the TraceID for effective sampling by itself. The other options that have been discussed are this option (in this PR, to attach a little additional randomness) and to specify a standard hashing function to use. Of those options, this option seems easiest and most reliable. 

Just saying "16 random bytes" doesn't say enough. If we say that every bit of the 128-bit TraceID is independently and identically random with p=0.5, then the random property of that ID we are trying to compute is the number of leading zeros. If we can assert that every bit is perfectly random, then we do not need this additional random variable to be propagated.

Author: I've written this OTEP under the assumption that we can assume nothing definite about TraceID randomness at the individual bit level. I believe it would take a W3C revision to do better than this. Although I am willing to continue updating this document with more possibilities if necessary, I think that's out of scope here.

Author: I've added to the caveats section this note about W3C and traceID randomness.

leading zeros using efficient instructions on modern computer
architectures.

For example, the value 3 means there were three leading zeros and
corresponds with being sampled at probabilities 1-in-1 through 1-in-8
but not at probabilities 1-in-16 and smaller.  Using one byte of
information we can convey a consistent sampling decision for sampling
rates as small as 2^-255.

### Proposed `tracestate` syntax

The consistent sampling decision and head trace sampling probability
will be propagated using four bytes of base16 content, as follows:

```

> tracestate: otelprob=PPRR

Reviewer: This does not match w3c spec (cf. https://www.w3.org/TR/trace-context/#combined-header-value), the expectation is that a vendor uses only a single key

> Example: `vendorname1=opaqueValue1,vendorname2=opaqueValue2`

We need to pick a vendor key for all of OTEL (`otel` or even just `ot`), and then decide on the encoding format for parts/attributes (which w3c does not prescribe)), for example:

```
tracestate: ot=p:PP;r:RR
tracestate: otel=pr:PPRR
```
Colons and semi-colons are allowed in values, but we need to consider if some future parts of state may need to contain (semi)colons.

Reviewer 2: Sounds like a plan. Shall we define _first_ in the Specification the way to populate `tracestate` ourselves @yurishkuro? Happy to do that, but I'd like if that's what you foresee.

Reviewer: @carlosalberto there are two independent things - how the tracestate format should look, and how SDK should interact with it. The latter is more complicated and I would say requires a prototype first, because there may be multiple touch points in the SDK that would need to mutate tracestate.

Reviewer 3: I agree with yuri that `tracestate: otelprob=PPRR` is not the intention of the w3c spec and that we should have a single key and define our own structure within the value. `ot` or `otel` both seem fine to me. Yuri's proposed semicolon-delimited value format `key1:value1;key2:value2` seems reasonable enough to me. I would probably prefer `pr:PPRR` to `p:PP;r:RR` because you need both keys in order for the value to be useful and the first option is shorter and makes it more clear the values are tightly coupled.

Author: @carlosalberto has a draft on this: https://github.com/open-telemetry/opentelemetry-specification/pull/1852

```

where `PP` are two bytes of base16 probability value and `RR` are two

> bytes of base16 random value.

Reviewer 3: If we limit to `2^-62` per @oertl suggestion, does it make sense still to use base 16? Base 10 is more human friendly and they both will take 2 characters to represent the max value so there is no space savings to use base 16.

Author: Yes, this is a good point. When @carlosalberto and I discussed this very point, we also considered that if these values were moved into the `traceparent` header, where they could really take just a byte or two, it would be more consistent to use base16 like the rest of the traceparent.

Author: I've added to the text that this proposal is about how we propagate, not about tracestate syntax. Whatever we decide for tracestate will dictate what we do here.

### Examples

The following `tracestate` value:

```
tracestate: otelprob=0a03
```

translates to

```
base16(probability) = 03 // 1-in-8 head probability
base16(randomness) = 0a // qualifies for 1-in-1024 sampling or greater
```

Any `TraceIDRatioBased` Sampler configured with probability 2^-10 or
greater will enable sampling this trace, whereas any
`TraceIDRatioBased` Sampler configured with probability 2^-11 or less
will stop sampling this trace.  The W3C `sampled` flag is set to true
when the probability value is less than or equal to the randomness
value.

## Internal details

The reasoning behind restricting the set of sampling rates is that it:

- Lowers the cost of propagating head sampling probability
- Limits the number of random bits required
- Makes math involving partial traces tractable.

[An algorithm for making statistical inference from partially-sampled
traces has been published](https://arxiv.org/pdf/2107.07703.pdf) that
explains how to work with a limited number of power-of-2 sampling rates.

### Behavior of the `TraceIDRatioBased` Sampler

The Sampler must be configured with a power-of-two probability

> `P=2^-S`.

Reviewer 3: Defined limit of `S<=62` per @oertl comment?

Author: I've changed the text to refer to "6 bits" of randomness, with 63 reserved to mean zero probability.

Using one byte to represent both the probability and randomness values
means each value is limited to 255.  As a special case, the head
probability `P=0` is represented using the probability value `S=255`,
meaning `P=0` is indistinguishable from `P=2^-255`.  We propose to
limit valid head sampling probabilities to `P=2^-254` or greater to
address this ambiguity.

If the context is a new root, the initial `tracestate` must be created
using geometrically-distributed random value `R` (as described above,
with maximum value 254) and the initial head probability value `S`.

If the context is not a new root, output a new `tracestate` with the
same `R` value as the parent context, using the Sampler's own value of

> `S` for the head probability.

Reviewer 3: Just to be clear, I am propagating tracestate with my own S value, not the value I received from the parent?

Author: Yes, this is what I mean by "the Sampler's own value of `S`". I'll try to improve this text.

Author: I've improved this text.

### Behavior of the `TraceIDRatioBased` Sampler

The Sampler must be configured with a power-of-two probability
`2**-S` except for the special case of zero probability, which is handled
specially.

If the context is a new root, the initial `tracestate` must be created
using geometrically-distributed random value `R` (as described above,
with maximum value 61) and the initial head probability value `S`.  If
the head probability is zero use `S=63`, the specified value for zero
probability.

If the context is not a new root, output a new `tracestate` with the
same `R` value as the parent context, and this Sampler's value of `S`
for the outgoing context's probability value (i.e., as the value for
`P`).

In both cases, set the `sampled` bit if `S<=R` and `S<63`.

### Behavior of the `ParentBased` sampler

The `ParentBased` sampler is unmodified by this proposal.  It honors
the W3C `sampled` flag and copies the incoming `tracestate` keys to
the child context.

### Behavior of the `AlwaysOn` Sampler

The `AlwaysOn` Sampler behaves the same as `TraceIDRatioBased` with `P=1` (i.e., `S=0`)

### Behavior of the `AlwaysOff` Sampler

> The `AlwaysOff` Sampler behaves the same as `TraceIDRatioBased` with `P=0` (i.e., `S=62`).

Reviewer 4: IIUC - This means selection probability is MAX VALUE vs. 0 (63). Is that a typo?

Author: There were a number of off-by-one errors stemming from changes following https://github.com/open-telemetry/opentelemetry-specification/pull/1899. Fixed.

## Worked example

The behavior of these tables can be verified by hand using a smaller
example.  The following table shows how these equations work where
`R`, `P`, and `S` are limited to 3 bits.

Values of `P`, which have the same encoded value and interpretation as
for the proposed `log_head_adjusted_count` field of OTEP 170, would be
interpreted as follows:

| `P` value | Adjusted count |
| -----     | -----          |
| 0         | Unknown        |
| 1         | 1              |
| 2         | 2              |
| 3         | 4              |
| 4         | 8              |
| 5         | 16             |
| 6         | 32             |
| 7         | 0              |

Note there are only 6 non-zero, non-unknown values for the adjusted
count. Thus there are six defined values of `R` and `S`.  The
following table shows `R` and the corresponding selection probability,
along with the calculated adjusted count for each `S`:

> | `R` value | `R` selection probability | `S=0` | `S=1` | `S=2` | `S=4` | `S=5` | `S=6` |

Reviewer 5: column `S=3` is missing

unbiased sampling decisions.  It is seen as a relatively difficult
task to define and specify a good enough hashing function, much less
to have it implemented in multiple languages.

Hashing is also computationally expensive. This proposal uses extra
data to avoid the computational cost of hashing TraceIDs.

### Restriction to power-of-two 

Restricting head sampling rates to powers of two does not limit tail
Samplers from using arbitrary probabilities.  The companion [OTEP
170](https://github.com/open-telemetry/oteps/pull/170) has discussed
the use of a `sampler.adjusted_count` attribute that would not be
limited to power-of-two values.  Discussion about how to represent the
effective adjusted count for tail-sampled Spans belongs in [OTEP
170](https://github.com/open-telemetry/oteps/pull/170), not this OTEP.

Restricting head sampling rates to powers of two does not limit
Samplers from using arbitrary effective probabilities over a period of
time.  For example, choosing 1/2 sampling half of the time and 1/4
sampling half of the time leads to an effective sampling rate of 3/8.

## Prior art and alternatives

Google's Dapper system propagated a field in its trace context called

> "inverse_probability", which is equivalent to adjusted count.  This

Reviewer 4: This may have been true at one point, but is not now.

Author: 🔪

head trace sampling probability associated with a context in order to
build span-to-metrics pipelines when the built-in `ParentBased`
Sampler is used.  Further motivation for supporting span-to-metrics
pipelines is presented in [OTEP
170](https://github.com/open-telemetry/oteps/pull/170).

A consistent trace sampling decision is one that can be carried out at
any node in a trace, which supports collecting partial traces.
OpenTelemetry specifies a built-in `TraceIDRatioBased` Sampler that
aims to accomplish this goal but was left incomplete (see a
[TODO](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased) 
in the v1.0 Trace specification).

We propose to propagate the necessary information alongside the [W3C
sampled flag](https://www.w3.org/TR/trace-context/#sampled-flag) using
`tracestate` with an `otel` vendor tag, which will require
(separately) [specifying how the OpenTelemetry project uses
`tracestate` itself](https://github.com/open-telemetry/opentelemetry-specification/pull/1852).

## Explanation

Two pieces of information are needed to convey consistent head trace
sampling probability:

> 1. The head trace sampling probability

Reviewer 6: nit: add a `.` (or remove the `.` from the 2nd item)

consistent sampling decision is accomplished by propagating a specific
random variable denoted `r`.  The random variable is a described by a
geometric distribution having shape parameter `1/2`, listed below:

| `r` Value | Selection Probability |
| ---------------- | --------------------- |
| 0 | 1/2 |
| 1 | 1/4 |
| 2 | 1/8 |
| 3 | 1/16 |
| ... | ... |
| 0 <= `r` <= 61 | 1/(2**(-`r`+1)) |
| ... | ... |
| 60 | 2**-61 |
| 61 | 2**-61 |

Such a random variable `r` can be generated using the following
pseudocode.

```golang
func nextRandomness() int {
  r := 0
  for r < 61 && nextRandomBit() == false {
    r++
  }
  return R
}
```

This can be computed from a stream of random bits as the number of
leadieng zeros using efficient instructions on modern computer
architectures.

For example, the value 3 means there were three leading zeros and
corresponds with being sampled at probabilities 1-in-1 through 1-in-8
but not at probabilities 1-in-16 and smaller.

### Proposed `tracestate` syntax

The consistent sampling randomness valuw (`r`) and and head sampling
probability value (`p`) will be propagated using two bytes of base16 content

> for each of the two fields, as follows,

Reviewer 6: ```suggestion
for each of the two fields, as follows:
```

```

> tracestate: otel=p:PP;r:RR

Reviewer 6: tracestate key value pairs might get [removed based on FIFO](https://www.w3.org/TR/trace-context/#tracestate-limits), do we have recommendation for OpenTelemetry SDKs such as "the OpenTelemetry SDKs are encouraged to preserve otel tracestate as much as possible, for example, always pop it to 1st key-value pair of tracestate" (which might be a bit off from the W3C TraceContext perspective).

```

where `PP` are two bytes of base16 probability value and `RR` are two
bytes of base16 random value.  These values are omitted when they are
unknown.

This proposal should be taken as a recommendation and will be modified
to [match whatever format OpenTelemtry specifies for its
`tracestate`](https://github.com/open-telemetry/opentelemetry-specification/pull/1852).
The choice of base16 encoding is therefore just a recommendation,
chosen because `traceparent` uses base16 encoding.

### Examples

The following `tracestate` value:

```
tracestate: otel=r:0a;p:03
```

translates to

```

> base16(probability) = 03 // 1-in-8 head probability

Reviewer 6: and the 2nd line is 1-in-512

| 3         | 1/16                      | 1     | 2     | 4     | 8     | 0     | 0     |
| 4         | 1/32                      | 1     | 2     | 4     | 8     | 16    | 0     |
| 5         | 1/32                      | 1     | 2     | 4     | 8     | 16    | 32    |

Notice that the sum of `r` selection probability times adjusted count
in each of the `s=*` columns equals 1.  For example, in the `s=5`
column we have `0*1/2 + 0*1/4 + 0*1/8 + 0*1/16 + 16*1/32 + 16*1/32 =
16/32 + 16/32 = 1`.  In the `s=2` column we have `0*1/2 + 0*1/4 +
4*1/8 + 4*1/16 + 4*1/32 + 4*1/32 = 4/8 + 4/16 + 4/32 + 4/32 = 1/2 +
1/4 + 1/8 + 1/8 = 1`.  We conclude that when `r` is chosen with the
given probabilities, any choice of `s` produces one expected span.

## Summary

The following table summarizes how the three Sampler cases behave with
respect to the incoming and outgoing values for `p`, `r`, and
`sampled`:

| Sampler                | Incoming `r` | Incoming `p` | Incoming `sampled` | Outgoing `r`    | Outgoing `p`   | Outgoing `sampled` |
| --                     | --           | --           | --                 | --              | --             | --                 |

> | Parent                 | unused       | expected     | respected          | passed through  | passed through | passed through     |

Reviewer 5: Parent-based sampler should not pass through inconsistent values of `r`, `p`, and `sampled` (see my comment below).

Author: I rewrote this as "checked and passed through", then added a table of range checks that should be applied, then documented the invariant as you stated it. I want there to be a special case for `sampled` is true and `p=63` to allow intentional sampling with zero adjusted count. I would be glad to prototype a composite sampler that applies either the probability sampler's decision (resulting in p in [1..62]) or a non-probability sampler's decision (effectively `p=63`, saying "collect this for some reason but don't count it") to show this in action.

| TraceIDRatio(Non-Root) | used         | unused       | ignored            | passed through  | set to `s+1`   | set to `p-1 < r+1` |
| TraceIDRatio(Root)     | n/a          | n/a          | n/a                | random variable | set to `s+1`   | set to `p-1 < r+1` |

There are several cases where the resulting span's
`log_head_adjusted_count` is unknown:

| Sampler                | Unknown condition |
| --                     | --                |
| Parent                 | no incoming `p`   |
| TraceIDRatio(Root)     | no incoming `r`   |
| TraceIDRatio(Non-Root) | none              |

There are cases where the combination of `p` and `r` and `sampled`
that cannot be generated by the built-in samplers.  The case where
sampled is true with `p=63` indicating 0% probability may be used when
recording spans that were selected by a different sampler while a
probability sampler is also in use.  These cases are known as "zero
adjusted count" contexts which are sampled with 0% probability.

The case where sampled is false with `p=1` indicating 100% probability
is an illogical condition.  See [Propagating `p` when
unsampled](#propagating-p-when-unsampled) below.

> 

Author: I prefer the first of your remedies, but want to refine it to accommodate the zero probability (`p=63`) as being allowed as a special case to allow sampling with zero adjusted count. Other than this special case, the outgoing `p` should be dropped, which is effectively 0 (unknown), when the invariant is broken. I've rewritten this section, please check. Thanks for pointing out how to simplify this!

head trace sampling probability associated with a context in order to
build span-to-metrics pipelines when the built-in `ParentBased`
Sampler is used.  Further motivation for supporting span-to-metrics
pipelines is presented in [OTEP
170](https://github.com/open-telemetry/oteps/pull/170).

A consistent trace sampling decision is one that can be carried out at
any node in a trace, which supports collecting partial traces.
OpenTelemetry specifies a built-in `TraceIDRatioBased` Sampler that
aims to accomplish this goal but was left incomplete (see a
[TODO](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased) 
in the v1.0 Trace specification).

We propose to propagate the necessary information alongside the [W3C
sampled flag](https://www.w3.org/TR/trace-context/#sampled-flag) using
`tracestate` with an `otel` vendor tag, which will require
(separately) [specifying how the OpenTelemetry project uses
`tracestate` itself](https://github.com/open-telemetry/opentelemetry-specification/pull/1852).

## Explanation

Two pieces of information are needed to convey consistent head trace
sampling probability:

> 1. The head trace sampling probability.
> 2. Source of consistent sampling decisions.

Reviewer: styling suggestion:
```suggestion
1. p-value representing the head trace sampling probability.
2. r-value representing the "randomness" as the source of consistent sampling decisions.
```
and refer to these as p-value and r-value instead of full names, because full names are misleading (p=2 is not a "probability value" by itself, but a specially encoded representation of such)

| 3                 | 1/4              |                        |
| ...               | ...              |                        |
| N                 | 2**(-N+1)        | 1 in 2**(N-1)          |
| ...               | ...              |                        |
| 61                | 2**-60           |                        |
| 62                | 2**-61           |                        |
| 63                | 0                | Maximum encoded value  |

[Described in OTEP
170](https://github.com/open-telemetry/oteps/pull/170), Span data
sampled by the `ParentBased` sampler will encode the value that was
propagated by the parent span as its "probability value" `p`.

The value `p=0` SHOULD NOT be propagated using `tracestate`
explicitly, because the equivalent interpretation can be obtained by
omitting `p`.

### Randomness value

With head trace sampling probabilities limited to powers of two, the
amount of randomness needed per trace context is limited.  A
consistent sampling decision is accomplished by propagating a specific
random variable denoted `r`.  The random variable is described by a
(truncated) geometric distribution having shape parameter `1/2`, listed below:

> | `r` Value        | Selection Probability | Sampling probability   |

Reviewer: I don't know how many people will be able to follow this. How about starting with a simpler explanation? Like generating a uniformly distributed 61-bit(ish) random value R such that a classic probabilistic sampler with probability P would sample if R <= P. That's the baseline that probably everyone will understand. Then we take into account that if P is discrete as 2**-s, then R can be similarly discreet, and its encoding can be reduced to counting the number of leading 0 bits (a couple concrete examples would probably help a lot).

The fact that r is a geometric distribution - is that relevant? Sure, we can mention it, after the more common sense explanation.

Finally, why do we need to list THAT algorithm (nextRandomness)? That's not how it will be implemented in practice in the SDKs. Why not show the more realistic algorithm that generates an n-bit random number and does bit-shifting to produce r-value?

Author: Thank you. I've updated this text with a more gentle introduction and removed some of the mathy words. I've changed the example code to use `rand.Int63()` and `bits.LeadingZeros64()` from the Golang standard library, also posted some code to compute it several other ways and compare the costs:

https://gist.github.com/jmacd/79c38c1056035c52f6fff7b7fc071274

Assuming an SDK uses at least 61 bits of *independent* randomness in the TraceID, it should be easy to compute in any case. The reason for this exposition is that not all TraceIDs will have independent randomness, and in that case all they need to compute an r-value, on average, are two random bits.

| ---------------- | --------------------- | ----                   |
| 0                | 1/2                   | 1-in-1                 |
| 1                | 1/4                   | 1-in-2 and above       |
| 2                | 1/8                   | 1-in-4 and above       |
| 3                | 1/16                  | 1-in-8 and above       |
| ...              | ...                   | ...                    |
| 0 <= `r` <= 60   | 1/(2**(-`r`-1))       | 1-in-2**-`r` and above |
| ...              | ...                   | ...                    |
| 58               | 2**-59                | 1-in-2**-58 and above  |
| 59               | 2**-60                | 1-in-2**-59 and above  |
| 60               | 2**-61                | 1-in-2**-60 and above  |
| 61               | 2**-61                | 1-in-2**-61 and above  |

Such a random variable `r` can be generated using the following
pseudocode.

```golang
func nextRandomness() int {
  r := 0
  for r < 61 && nextRandomBit() == false {
    r++
  }
  return R
}
```

This can be computed from a stream of random bits as the number of

> leading zeros using efficient instructions on modern computer

Reviewer: > leading zeros

or trailing zeros

Author: Included in the gist.

`tracestate`](https://github.com/open-telemetry/opentelemetry-specification/pull/1852).
The choice of base16 encoding is therefore just a recommendation,
chosen because `traceparent` uses base16 encoding.

### Examples

The following `tracestate` value:

```
tracestate: otel=r:0a;p:03
```

translates to

```
base16(probability) = 03 // 1-in-4 head probability
base16(randomness) = 0a // qualifies for 1-in-1024 or greater probability consistent sampling
```

Any `TraceIDRatioBased` Sampler configured with probability 2**-9 or
greater will enable sampling this trace, whereas any
`TraceIDRatioBased` Sampler configured with probability 2**-10 or less
will stop sampling this trace.

## Internal details

The reasoning behind restricting the set of sampling rates is that it:

- Lowers the cost of propagating head sampling probability
- Limits the number of random bits required
- Avoids floating-point to integer rounding errors
- Makes math involving partial traces tractable.

[An algorithm for making statistical inference from partially-sampled
traces has been published](https://arxiv.org/pdf/2107.07703.pdf) that
explains how to work with a limited number of power-of-2 sampling rates.

> ### Behavior of the `TraceIDRatioBased` Sampler

Reviewer: I would recommend introducing a new sampler with a different name and not touching the TraceIDRatioBased at all, because the actual mechanism has nothing to do with "trace-id ratio" anymore.

Author: What would you propose that we do with `TraceIDRatioBased`, then? It has TODOs in the specification and the implementations we have rely on unspecified use of TraceID bits. If the best we can do is not touch the `TraceIDRatioBased` logic, then I propose we name it to `IncosistentProbabilityBased` and leave it as deprecated with the existing caveat that it should only be used at the root.

From the perspective of this OTEP, traces rooted in `InconsistentProbabilityBased` samplers would not be able to convey head probability.

From the perspective of OTEP 170, all spans in such a trace would have to record unknown adjusted count. We could _in the future_ resume discussions about a `sampler.adjusted_count` attribute to allow arbitrary sampling probability at the root span while propagating unknown head probability. Consumers could then use `InconsistentProbabilityBased` samplers w/ arbitrary probability at the root of their trace, perform trace assembly, and then perform span-to-metrics over sampled traces.

As for naming the proposed new Sampler, @oertl how do you feel about `ConsistentProbabilityBased`? 

(Actually, do we need all these "Based" strings anyway, could we just talk about "Parent", "TraceIDRatio", "ConsisterntProbability", ...?)

Reviewer: > What would you propose that we do with TraceIDRatioBased, then

Deprecated it? Since the spec never said how it needs to be implemented, do existing SDKs even have this sampler today? And if they do, how can it be consistent across langs?

Author: OK. Yes the SDKs have implemented this Sampler because among the builtin samplers, it's the only one useful as a root other than always-on and always-off. It's just not specified in a consistent way. That's why I proposed it _could_ be kept as a sampler for non-power-of-two-probabilities, which the legacy TraceIDRatioBased sampler supports, but I think it would be better to replace it and round the input sampling probabilities to powers-of-two. It's fine to choose a new name, my goal is that we end up with just one behavior.

Could we just call this single new sampler the ProbabilitySampler? Could we merge this OTEP and debate names and whether to deprecate the legacy behavior in another PR?

Reviewer: I would prefer more special name like "ConsistentSampler" over "ProbabilitySampler", as the letter is pretty common in pre-OTEL SDKs and means "flip a p-weighted coin"

condition and follow the incoming context's `sampled` flag.

The span's `log_head_adjusted_count` field is set to the outgoing `p`
unless `r` is unknown, in which case it MUST be set to zero (unknown
probability).

### Behavior of the `ParentBased` sampler

The `ParentBased` sampler is modified by this proposal.  It honors
the W3C `sampled` flag and copies the incoming `tracestate` keys to
the child context.

The span's `log_head_adjusted_count` field is set to the incoming
value of `p` when both `p` and `r` are defined.  When `r` is not
defined, the span's `log_head_adjusted_count` MUST be set to 0
indicating unknown probability, because the decision cannot be made
consistently across the trace.

### Behavior of the `AlwaysOn` Sampler

The `AlwaysOn` Sampler behaves the same as `TraceIDRatioBased` with
100% sampling probability (i.e., `p=1`).

### Behavior of the `AlwaysOff` Sampler

The `AlwaysOff` Sampler behaves the same as `TraceIDRatioBased` with
zero probability (i.e., `p=63`).

## Worked 3-bit example

The behavior of these tables can be verified by hand using a smaller
example.  The following table shows how these equations work where

> `r`, `p`, and `s` are limited to 3 bits.

Reviewer: ```suggestion
`r`, `p`, and `s` are limited to 3 bits instead of 6 bits.
```

Values of `p`, which have the same encoded value and interpretation as
for the proposed `log_head_adjusted_count` field of OTEP 170, would be
interpreted as follows:

| `p` value | Adjusted count |
| -----     | -----          |
| 0         | Unknown        |
| 1         | 1              |
| 2         | 2              |
| 3         | 4              |
| 4         | 8              |
| 5         | 16             |
| 6         | 32             |
| 7         | 0              |

Note there are only 6 non-zero, non-unknown values for the adjusted
count. Thus there are six defined values of `r` and `s`.  The

> following table shows `r` and the corresponding selection probability,

Reviewer: Why do we care about "`r` selection probability"? What is the objective - to generate r-value or to sample the trace?

Author: I think it's appropriate in an OTEP to explain technical details, they don't need to go into the specification. The reason we want to see the probability of each r-value is that it helps convince ourselves that the math works out. The r-value selection probabilities are being multiplied by adjusted counts in the example that is being worked out, that is why they are here.

Above, I've explained that an SDK is not required to use independent random bits in its TraceID. r-values can be calculated in any way that generates this distribution.

| --              | --                               |
| `p < 0`         | drop `p` from tracestate         |
| `p > 63`        | drop `p` from tracestate         |
| `r < 0`         | drop `r` and `p` from tracestate |
| `r > 61`        | drop `r` and `p` from tracestate |

There are cases where the combination of `p` and `r` and `sampled` are
inconsistent with each other.  The `sampled` flag is equivalent to the
expression `p - 1 <= r`.  When the invariant `sampled <=> p - 1 <= r`
is violated, the `ParentBased` sampler MUST correct the propagated
values as discussed below.

The violation is always addressed by honoring the `sampled` flag and
setting `log_head_adjusted_count` to either 0 (Unknown) or 63 (Zero).

If `sampled` is false and the invariant is bilated, drop `p` from the
outgoing context to convey unknown head probability.  Set
`log_head_adjusted_count` to 0.

The case where `sampled` is true with `p=63` indicating 0% probability
may by regarded as a special case to allow zero adjusted count
sampling, which permits non-probabilistic sampling to take place in
the presence of probability sampling.  Set `log_head_adjusted_count`

> to 63.

Author: In words, I think we should support a sampling policy like "Sample at 3.125% AND at least 1 per hour". A probability sampler will compute the 3.125% sample (i.e., 2^-5 or `p=6`) and a non-probability sampler will compute the one per hour sample with `p=0`. We can spell out rules for composing these, and this is why I have referred to the important special case of zero adjusted count. The probability sampler outputs an adjusted count of zero for any span it does not select. If another sampler decides to sample, we have the important special case of a sampled trace with zero adjusted count -- i.e., a non-probabilistic sample.

Stepping back, there is a higher-level invariant guiding this discussion, one stating that the sum of adjusted counts equals the true number of spans in the population. The `sampled <=> p-1<=r` invariant only needs to be applied when `p<63` because when `p=63` we have by definition broken the higher-level invariant: any time you intentionally choose zero-probability sampling you break the higher-level invariant, so we may as well exclude the lower-level invariant at that point. Thus, the special case of `sampled=true` and `p=63`. We would still want to have `r` defined in that case, because a child span somewhere in the trace could change the decision and therefore begin using a non-zero adjusted count for a sub-trace.

> to 63.

Author: @yurishkuro and @pavolloffay please confirm that a policy like "Sample at 3.125% AND at least 1 per hour"  is a realistic sampler configuration in today's Jaeger sampler? This is a composite sampler made up of a probability sampler and a rate-limited sampler, for example.

> to 63.

Reviewer: > please confirm that a policy like "Sample at 3.125% AND at least 1 per hour" is a realistic sampler configuration in today's Jaeger sampler

Confirmed. The "at least 1 per hour" part is useful when using adaptive sampling backend which works best when it's aware of all endpoints in the architecture, including those with low QPS that may otherwise never be sampled by the probability sampler (because there can be several orders of magnitude difference in QPS so the conservative default sampling probability has to be very low).

> to 63.

Author: The rule for composing more than one `log_head_adjusted_count` value will say that it is OK to combine any number of Unknown p-values with a single known p-value and to `OR` their `sampled` flags together.

> to 63.

Author: We can recommend the use of adaptive probability sampling and we've shown a couple of examples of this in action, though we know that it's impossible to implement a hard rate limit without introducing bias.

My proposal allows easily combining non-probabilistic samplers (which all produce unknowns) with a single probabilistic sampler so that adjusted counts are always known, with no modification to the other samplers. It may be better to use an adaptive strategy, but it's a very different approach.

> switches once in an hour to a sampling rate of 100%.

This is true, but it doesn't give me a good feeling. Faced with non-probabilistic Samplers, we can either force them to become probabilistic or accommodate known zeros. I think that both are viable, but it is more difficult to reason about the data produced by an adaptive sampler, and if we're not careful there may be temporal or systematic bias produced. It's out-of-scope go into a detailed analysis of this here, but merely the fact that sophisticated analysis is required suggests that some users will better off with a simple approach.

> kind of unnatural anyway

This may just be me, but it feels natural to use zero for this application. We can add any number of zeros to a sum without changing the expected value.

I agree that there is something unnatural about sampling at 0%, because it is not possible to preserve an expected value of the sum when doing so. It is biased by definition. But for the application at hand, we are sampling with a non-zero probability and the use of zero meaningfully conveys that decision in a way that allows non-probabilistic samplers to exist. I'm not against adaptive sampling, it's just more complex and I'd like to offer both options.

> to 63.

Reviewer 5: @jmacd

> combining non-probabilistic samplers (which all produce unknowns)

I am still trying to understand that. Does combining multiple samplers (one probabilistic and multiple non-probabilistic samplers) mean applying them all to the same span? If yes, is it possible that multiple samplers decide to sample the same span at the same time? If so, which value is propagated for `p`? The propabilistic sampler would set `p<63` while the non-probabilistic sampler would set `p=63`.

> to 63.

Author: > while the non-probabilistic sampler would set p=63.

**The non-probabilistic sampler would set `p=0` for unknown.**

The composite sampler will see a number of unknowns `p=0` and a single known `p>0`. The resulting `p` is the known value (63 or less than 63, just not 0), while `sampled` may be true or false.

In other words, as long as we have one probability sampler we always know the adjusted count. We can combine the one probability sampler with any number of non-probability samplers.

> to 63.

Reviewer 5: I am not talking about the adjusted count, but what is propagated in the `tracestate`. Non-probabilistic samplers are required to set `sampled = true` and `p = 63` (sampling rate = 0) in order to ensure that all subsequent parent-based samplers also report zero adjusted counts while probabilistic samplers are required to set `sampled = true` and `1 <= p < 63` (sampling rate > 0). As there is only one `tracestate` for the same span, there is a conflict. It must be somehow ensured that non-probabilistic samplers do never overwrite the `tracestate` of the probabilistic samplers if they are applied to the same span.

> to 63.

Reviewer 5: The only reason for zero adjusted counts that I can see is the parallel use of multiple samplers applied to the same spans. In this case it must be ensured that only one of the samplers delivers non-zero adjusted count values in order to get unbiased results. 

@jmacd
> I think that both are viable, but it is more difficult to reason about the data produced by an adaptive sampler, and if we're not careful there may be temporal or systematic bias produced. It's out-of-scope go into a detailed analysis of this here, but merely the fact that sophisticated analysis is required suggests that some users will better off with a simple approach.

The simultaneous application of multiple samplers to the same span makes things much more complicated (see also my previous  comment) than a single sampler that chooses the sampling rate dynamically. Any sampler implementation that picks some nonzero sampling rate (possbily different for every single span and dependent on some internal logic) and makes the sampling decision according to `sampled <=> p-1 <= r` will be unbiased by definition. Actually, this allows a lot of freedom how to sample. (See my prototype implementations. Any implementation of [AbstractConsistentSampler](https://github.com/dynatrace-research/opentelemetry-sampling-poc/blob/main/poc/src/main/java/com/dynatrace/research/otelsampling/sampling/AbstractConsistentSampler.java) will be unbiased. Furthermore, different samplers can be easily composed as demonstrated by the [ComposedSampler](https://github.com/dynatrace-research/opentelemetry-sampling-poc/blob/main/poc/src/main/java/com/dynatrace/research/otelsampling/sampling/ComposedSampler.java)). Reasoning about the data is easier. There is the advantage that adjusted counts will be always greater than or equal  to 1 and that there will never be any duplicates. In the presence of multiple samplers, the same span might be sampled more than once, once with a non-zero adjusted count and the other times with a zero adjusted count.

I think using multiple samplers in parallel to sample the same span should be generally avoided.

> to 63.

Author: > The only reason for zero adjusted counts that I can see is the parallel use of multiple samplers applied to the same spans. In this case it must be ensured that only one of the samplers delivers non-zero adjusted count values in order to get unbiased results.

You wrote "In this case it must be ensured that only one of the samplers delivers **non-zero** adjusted count values in order to get unbiased results." (emphasis mine) but this is NOT what I'm describing. I am saying "it must be ensured that only one of the samplers is a probability sampler."

I am using the expression "non-probability sampler" to describe a decision that is not based in probability and therefore **MUST have Unknown ​probability**.  The only composition I have described is a single probability sampler **having a single Known probability** combined with any number of non-probability samplers. The combination allows for any number of Unknowns and a single Known. 

We are not specifying how to multiple probability samplers in parallel.  The `log_head_adjusted_count` field specified in OTEP 170 is specified as a sample of all spans. As it covers all spans in "the" population, there could not be a need for more than one of them.

If a user did want to use multiple probability samplers in parallel, then I would presume: (1) they are independent, (2) they cover sub-populations of spans. In an early draft of OTEP 170 I spelled out how this might look, e.g., with span attributes like `sampler.name.adjusted_count` for Sampler named `name`. This could be specified later, and when that time comes it will become natural to set `log_head_adjusted_count=63` because there are multiple parallel probability samplers.

For now, the specified reason to use `log_head_adjusted_count=63` is that you have an always-off sampler.

I will update this PR with more information on composing Known and Unknown probability values, since there is an immediate use for combining probability and non-probability samplers

> to 63.

Reviewer 5: @jmacd I finally figured out a use case which justifies to break the invariant `sampled <=> p-1 <= r`. Maybe this is exactly what you meant. Basically, if `sampled`, `p`, and `r` are all known, the invariant `sampled <=> p-1 <= r` should be never violated when propagating them to child spans.

However, a valid use case is, for example, to sample up to 100 spans every minute. This can be achieved with a reservoir sampling buffer of size 100 that is reset every minute. It is possible to do reservoir sampling in a consistent way (see [ReservoirSampler.java](https://github.com/dynatrace-research/opentelemetry-sampling-poc/blob/main/poc/src/main/java/com/dynatrace/research/otelsampling/sampling/ReservoirSampler.java)). The problem is that the sampling decision is not known at propagation time. Only at the end of every minute, it is known which spans survived in the reservoir and got finally sampled. This means that neither `sampled` nor `p` can be correctly set and propagated. Therefore, reservoir sampling is incompatible with parent-based sampling which requires `sampled` and `p` of the parent to be known. (Reservoir sampling is compatible with all other consistent built-in samplers that do not depend on the parent sampling decision.) Since `sampled` is not an optional field, `sampled=unknown` cannot be directly expressed. A workaround would be to set  `sampled=true` together with `p=63` (sampling probability = 0%) which is a contradiction and violates the invariant, but can be used to express the special situation where both `sampled` and `p` are not known at propagation time.

I have the feeling that we are mixing up `p` and `log_head_adjusted_count` in our discussion. As far as I understand, `p` is the value that is propagated to child spans, while `log_head_adjusted_count` describes the adjusted count of sampled spans. The parent-based sampler uses the incoming `p` to set `log_head_adjusted_count=p`, if the span gets sampled. I am still not aware of any use case for reporting spans with `log_head_adjusted_count=63` (which corresponds to adjusted count = 0). I can imagine cases for `log_head_adjusted_count=0` (which corresponds to adjusted count = unknown). For example, when a parent-based sampler did not receive a valid `p` value while `sampled=true`.

(The only situation I can imagine where adjusted count = 0 could be useful is when multiple samplers are simultaneously applied to the same span. In this case only one primary sampler is allowed to report nonzero adjusted counts while all others have to set adjusted count = 0 for their sampled spans to avoid bias. However, I think you agreed that using multiple samplers in parallel does not make much sense.)

> to 63.

Author: This thread is too long for anyone to read. :-)

I think we're actually agreeing about things and have become befuddled with our own terminology, so I've updated the text of recently-merged #170 with detail to help clarify this topic. The new PR #175 contains a table that specifies how to compose samplers.

I realized I never responded to some of the `AbstractConsistentSampler` and `ComposedSampler` above. In the new table in #175 I've written that to combine two consistent samplers, take the `min()` p-value (as you stated), but it also covers the case of combining known and unknown values as I've stressed it should.

If `sampled` is true with `p<63`, drop `p` from the outgoing context
to convey unknown head probability.  Set `log_head_adjusted_count` to
0.

## Prototype

[This proposal has been prototyped in the OTel-Go
SDK.](https://github.com/open-telemetry/opentelemetry-go/pull/2177) No
changes in the OTel-Go Tracing SDK's `Sampler` or `tracestate` APIs
were needed.

## Trade-offs and mitigations

### Not using TraceID randomness

It would be possible, if TraceID were specified to have at least 61
uniform random bits, to compute the randomness value described above
as the number of leading zeros among those 61 random bits.

> This proposal requires modifying the W3C traceparent specification,

Reviewer: ```suggestion
However, this would require modifying the W3C traceparent specification,
```

therefore we do not propose to use bits of the TraceID.

> [This issue has been filed with the W3C trace context group.](https://github.com/w3c/trace-context/issues/463)

Reviewer: the linked issue is not about guaranteeing randomness of TraceID.

follows [research by Otmar Ertl](https://arxiv.org/pdf/2107.07703.pdf).

### p-value

To limit the cost of this extension and for statistical reasons
documented below, we propose to limit head trace sampling probability
to powers of two.  This limits the available head trace sampling
probabilities to 1/2, 1/4, 1/8, and so on.  We can compactly encode
these probabilities as small integer values using the base-2 logarithm
of the adjusted count.

Using six bits of information we can convey unknown and known sampling
rates as small as 2**-61.  The value 63 is reserved to mean sampling
with probability 0, which conveys an adjusted count of 0 for the
associated context.

When propagated, the "p-value" as it is known will be interpreted as
shown in the folowing table.  The p-value for known sampling
probabilities is the negative base-2 logarithm of the probability,
offset by +1 to so that the 0 p-value can be treated as unknown (for
backwards compatibility):

| p-value | Head Probability | Note                                                 |
| -----   | -----------      | ----                                                 |

> | 0       | Unknown          | Do not propagate `p=0`, instead omit from tracestate |

Author: Non-blocking comment for @jsuereth: Note there is a question about field presence here. If OTel adopts the use of proto3 field presence then we can distinguish between 0 and "not present", that would change (and simplify) the backwards compatibility story. Then p-values would equal the effective `s` value and the invariant would be written `sampled <=> p <= r`.

| 1                | 1/4                      | 1/2 and above                  |
| 2                | 1/8                      | 1/4 and above                  |
| 3                | 1/16                     | 1/8 and above                  |
| ...              | ...                      | ...                            |
| 0 <= `r` <= 60   | 1/(2**(-`r`-1))          | 2**-`r` and above              |
| ...              | ...                      | ...                            |
| 58               | 2**-59                   | 2**-58 and above               |
| 59               | 2**-60                   | 2**-59 and above               |
| 60               | 2**-61                   | 2**-60 and above               |
| 61               | 2**-61                   | 2**-61 and above               |

Such a random variable `r` can be generated using efficient
instructions on modern computer architectures, for example:

```golang
import (
	"math/rand"
	"math/bits"
)

func nextRValueLeading() int {
	x := uint64(rand.Int63()) // 63 least-significant bits are random
	y := x << 1 | 0x7         // 61 most-significant bits are random
	return bits.LeadingZeros64(y)

> }

Reviewer: alternative (may be off-by-one), which is language and std-lib neutral:
```go
const topbit = uint64(1) << 61
x := uint64(rand.Int63())
for n := 0; n < 61; n++ {
    if x & topbit == topbit {
        return n
    }
    x = x << 1
}
return 61
```

> }

Reviewer: alternative which is language & std-lib neutral (may be off by one):
```go
x := uint64(rand.Int63())
for r := 0; r < 61; r++ {
    if x & 0x1 == 0x1 {
        return r
    }
}
return 61
```

Author: I added this as a second example, computing the number of trailing zeros. I added it to the gist and verified that it computes the same result as the version using `bits.TrailingZeros64()`.

| 2                | 1/8                      | 1/4 and above                  |
| 3                | 1/16                     | 1/8 and above                  |
| ...              | ...                      | ...                            |
| 0 <= `r` <= 60   | 1/(2**(-`r`-1))          | 2**-`r` and above              |
| ...              | ...                      | ...                            |
| 58               | 2**-59                   | 2**-58 and above               |
| 59               | 2**-60                   | 2**-59 and above               |
| 60               | 2**-61                   | 2**-60 and above               |
| 61               | 2**-61                   | 2**-61 and above               |

Such a random variable `r` can be generated using efficient
instructions on modern computer architectures, for example we may
compute the number of leading zeros using hardware support:

```golang
import (
	"math/rand"
	"math/bits"
)

func nextRValueLeading() int {
	x := uint64(rand.Int63()) // 63 least-significant bits are random
	y := x << 1 | 0x7         // 61 most-significant bits are random
	return bits.LeadingZeros64(y)
}
```

> Or we may compute the number of trialing zeros, for example:

Reviewer: ```suggestion
Or we may compute the number of trailing zeros instead, for example:
```

```golang
import (
	"math/rand"
)

> func nextRValueLeading() int {

Reviewer: ```suggestion
func nextRValueTrailing() int {
```

| p-value | Head Probability | Note                                                 |
| -----   | -----------      | ----                                                 |
| 0       | Unknown          | Do not propagate `p=0`, instead omit from tracestate |
| 1       | 1                |                                                      |
| 2       | 1/2              |                                                      |
| 3       | 1/4              |                                                      |
| ...     | ...              |                                                      |
| N       | 2**(-N+1)        | 1 in 2**(N-1)                                        |
| ...     | ...              |                                                      |
| 61      | 2**-60           |                                                      |
| 62      | 2**-61           |                                                      |
| 63      | 0                | Maximum encoded value                                |

[Described in OTEP
170](https://github.com/open-telemetry/oteps/pull/170), the
`ParentBased` sampler will use the incoming context's p-value as

> specified here to set the span's `log_head_adjusted_count` field.

Author: As you note, if the `tracestate` contains this information (and more) then why would we store it in a field? My hope is that this use of `tracestate` demonstrates value such that W3C would accept a version-1 `traceparent` that includes the same information for 1-3 extra bytes, depending on the specifics. If W3C would also specify that 62 specific bits of the TraceID are required to be uniform random, that alleviates the need for an `r` variable, which means only requiring the `p` variable _which is only required when sampled_ so the total cost w/ a version-1 `traceparent` is very low--that's why `log_head_adjusted_count` is a dedicated field.

> specified here to set the span's `log_head_adjusted_count` field.

Reviewer 7: We already have tracestate so I see no point in adding the new field, that is what I am trying to say. Unless you say that the tracestate for this entry should not be sent then I see no point in having the new field

> specified here to set the span's `log_head_adjusted_count` field.

Reviewer 7: https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L89

> specified here to set the span's `log_head_adjusted_count` field.

Author: Fixed.

instructions on modern computer architectures, for example we may
compute the number of leading zeros using hardware support:

```golang
import (
	"math/rand"
	"math/bits"
)

func nextRValueLeading() int {
	x := uint64(rand.Int63()) // 63 least-significant bits are random
	y := x << 1 | 0x7         // 61 most-significant bits are random
	return bits.LeadingZeros64(y)
}
```

Or we may compute the number of trailing zeros instead, for example:

```golang
import (
	"math/rand"
)

func nextRValueTrailing() int {
	x := uint64(rand.Int63())
	for r := 0; r < 61; r++ {
		if x & 0x1 == 0x1 {
			return r
		}
		x = x >> 1
	}
	return 61
}
```

More examples for calculating r-values are shown in
[here](https://gist.github.com/jmacd/79c38c1056035c52f6fff7b7fc071274).
For example, the value 3 means there were three leading zeros and
corresponds with being sampled at probabilities 1-in-1 through 1-in-8
but not at probabilities 1-in-16 and smaller.

### Proposed `tracestate` syntax

The consistent sampling r-value (`r`) and and head sampling
probability p-value (`p`) will be propagated using two bytes of base16
content for each of the two fields, as follows:

```

> tracestate: otel=p:PP;r:RR

Author: Stated below "proposal should be taken as a recommendation". After https://github.com/open-telemetry/opentelemetry-specification/pull/1852 makes progress we can update our OTEPs if that will help.

```

where `PP` are two bytes of base16 p-value and `RR` are two bytes of

> base16 r-value.  These values are omitted when they are unknown.

Author: Again, this is just speculative recommendation. Let's put this question in https://github.com/open-telemetry/opentelemetry-specification/pull/1852. 

The W3C uses two bytes for "flags" despite having only 1 bit in use, so I imagined following their precedent and using two base16 bytes. Ideally, the version-1 W3C `traceparent` TraceID has 62 well-specified uniform random bits and then the additional `p` value can be had for as little as 1 byte in most cases (i.e., unknown probability and 100% sampling through 2**-14 sampling would require only one byte).

> base16 r-value.  These values are omitted when they are unknown.

Reviewer 7: Please make sure the comment is capture there.

[An algorithm for making statistical inference from partially-sampled
traces has been published](https://arxiv.org/pdf/2107.07703.pdf) that
explains how to work with a limited number of power-of-2 sampling rates.

### Behavior of the `TraceIDRatioBased` Sampler

The Sampler MUST be configured with a power-of-two probability
expressed as `2**-s` with s being an integer in the range [0, 61]
except for the special case of zero probability.

If the context is a new root, the initial `tracestate` must be created
with randomness value `r`, as described above, in the range [0, 61].
If the context is not a new root, output a new `tracestate` with the
same `r` value as the parent context.

When sampled, in both cases, the context's p-value `p` is set to the
value of `s+1` in the range [1, 62].  If the sampling probability is
zero (the special case where `s` is undefined), use `p=63` the
specified value for zero probability.

In both cases, set the sampled bit if the outgoing `p` minus one is
less than or equal to the outgoing `r` (i.e., `p-1 <= r`).

If the context is not a new root and the incoming context's r-value
is not set, the implementation SHOULD notify the user of an error

> condition and follow the incoming context's `sampled` flag.

Reviewer 7: This statement makes this feature "required" for all deployments, which I think it should not be.

Author: This level of detail about specific error handling, I think, can be debated in https://github.com/open-telemetry/opentelemetry-specification/pull/1899.

Reviewer 7: Please make sure the comment is capture there.

(i.e., not power-of-two) probabilities.  Because of this, we may keep
the current (under-specified) `TraceIDRatioBased` sampler and rename
it `ProbabilitySampler` to convey that it does behave in a specified
way with respect to the bits of the TraceID.

### Not using TraceID randomness

It would be possible, if TraceID were specified to have at least 61
uniform random bits, to compute the randomness value described above
as the number of leading zeros among those 61 random bits.

However, this would require modifying the W3C traceparent specification,
therefore we do not propose to use bits of the TraceID.

[This issue has been filed with the W3C trace context group.](https://github.com/w3c/trace-context/issues/463)

### Not using TraceID hashing

It would be possible to make a consistent sampling decision by hashing
the TraceID, but we feel such an approach is not sufficient for making
unbiased sampling decisions.  It is seen as a relatively difficult
task to define and specify a good enough hashing function, much less
to have it implemented in multiple languages.

Hashing is also computationally expensive. This proposal uses extra

> data to avoid the computational cost of hashing TraceIDs.

Reviewer 7: I would disagree with this argument, it is more expensive to parse the strings from the tracestate, communicate them over network, etc. than a hash of a 16-byte array (at least this is my intuition).

Author: @oertl has defended this point in the past with a link to https://github.com/rurban/smhasher, which shows how expensive and failure-prone a number of hashing functions are. My position is that the math of computing adjusted counts from inverse probabilities is likely to have artifacts as a result of low-quality hashing. The better long-term solution is a W3C `traceparent` with 62 specified uniform random bits.

### Restriction to power-of-two

Restricting head sampling rates to powers of two does not limit tail
Samplers from using arbitrary probabilities.  The companion [OTEP
170](https://github.com/open-telemetry/oteps/blob/main/text/trace/0170-sampling-probability.md) has discussed
the use of a `sampler.adjusted_count` attribute that would not be
limited to power-of-two values.  Discussion about how to represent the
effective adjusted count for tail-sampled Spans belongs in [OTEP
170](https://github.com/open-telemetry/oteps/blob/main/text/trace/0170-sampling-probability.md), not this OTEP.

Restricting head sampling rates to powers of two does not limit
Samplers from using arbitrary effective probabilities over a period of
time.  For example, choosing 1/2 sampling half of the time and 1/4

> sampling half of the time leads to an effective sampling rate of 3/8.

Reviewer: It may be better to pick a different example. p=3/8 looks neat, but what if I wanted p=7/16? I would need to pick p=1/2 vs. p=1/4 in proportion 3:1
```
1/2 * 0.75 + 1/4 * 0.25 = 7/16
```

Reviewer 8: Perhaps an even better example would be to show that a more traditional 5% sampling rate can be obtained by sampling with 1/16 probability 60% of the time and with 1/32 probability 40% of the time.

Author: Applied.

Author: Done!
